### PR TITLE
Basic Support for scale.* properties

### LINKED ISSUE #2: Enumerate Scale Properties
### Background

- [x] Look at description and changes of #27 to see the infrastructure for adding nested property (`bin.maxbins`) -- note that I might miss something in the description, but if that's the case, you'll notice problem as you debug.

### 1st step Scale.type
- [x] add `scale.type` (one PR) 
  - [x] understand what `scale.type` means from Vega-Lite docs
  - [x] Add stuff like in #27
  - spec constraints (add to `spec.ts`)
    - [x] `omitBarAreaForLogScale` -- don't use `bar` and `area` mark for log scale. 
  - encoding constraints
    - [x] `dataTypeMatchesScaleType` -- look at 
    - [x] `omitBinForLogScale` (originally https://github.com/vega/compass/issues/151) 
  - [x] Add Example Query to `examples/`
  - [x] add test for `enumerate`, 
  - [x] add test for`generate` 
  - [x] add test for all new constraints

### Scale.*
Repeat the process for other scale properties (one PR for each)

- add ones that are required by other tasks

  - [ ] `type`
       - `clamp`:  Q, T
       - `exponent`: pow
       - `round`: Q, T 
          - accept types of values depending on scale type 
  - [x] `zero`  --> zero doesn't play well with [ ScaleType.ORDINAL, LOG, TIME, UTC]. I don't think I'm missing anything else...
    - https://github.com/uwdata/compassql/issues/105
  - [x] `bandSize`
    - [ ] #93
    - ~~bandSize must be at least 0~~
  - [x] `range`
    - [ ] #101  
    - ~~values must contain two or more values.~~
  - [x] `domain`
  - [x] `round`
  - [x] `clamp`
    - must have continuous domain / continuous domain (quantitative and time types only)
  - [x] `nice`
    - similar to clamp.. quantitatiev and time. 
  - [x] `exponent`
  - [x] `useRawDomain`

--- LATER ---

  - [ ] `padding`
    - works with channel.x, channel.y --> uses pixels
    - ???  padding (0, 1) for rangeBands  ???  -- LATER

### PR DESCRIPTION
**Pull request by felixcodes**
Fix #2. 
Add test and basic support for scale.clamp, scale.domain, scale.exponent, scale.range, scale.round

### CODE REVIEW

**Comment by kanitw** [comment]
How do I know that you can successfully enumerate round.  (There is no such test yet.)

File: src/property.ts [hunk 1]
```diff
@@ -90,7 +93,8 @@ export const DEFAULT_PROPERTY_PRECEDENCE: Property[] =  [
 
   // Nested Encoding Property
   Property.SCALE_TYPE,
-  Property.SCALE_ZERO
+  Property.SCALE_ZERO,
+  Property.SCALE_ROUND,
```
Comment (kanitw) [inline src/property.ts]: Why Round after ZERO here?  

Unless you have convincing reason to do so, always use alphabetical order for readability.  (You actually do it else where in this PR)

File: src/property.ts [hunk 2]
```diff
@@ -114,6 +118,11 @@ export const NESTED_ENCODING_PROPERTIES: NestedEncodingProperty[] = [
     property: Property.SCALE_ZERO,
     parent: 'scale',
     child: 'zero'
+  },
+  {
```
Comment (kanitw) [inline src/property.ts]: alphabetical order

**Comment by kanitw** [comment]
Rebase this to fork from #141 because i will merge that first

(In the future, if you see conflicts coming due to file splitting -- you have to fork from your split branch and say in a PR that we have to merge split branch before a follow-up branch)

**Comment by felixcodes** [comment]
Rebased onto #141.

**Comment by kanitw** [comment]
Still have conflict due to merge squash, you might want to **interactive rebase** to master and drop all commits from #141 since they are already squashed.  (After force push, this PR should contain relevant commits.

**Comment by kanitw** [comment]
Please mark this as WIP until it's ready

File: src/config.ts [hunk 3]
```diff
@@ -40,8 +40,12 @@ export interface QueryConfig {
 
   scaleClamps?: boolean[];
 
+  scaleDomains?: Array<string | number[] | string[]>;
```
Comment (kanitw) [inline src/config.ts]: domain shouldn't be a string?

File: src/model.ts [hunk 4]
```diff
@@ -57,6 +57,15 @@ export interface EnumSpecIndex {
   /** List of indice tuple for encoding mappings that require enumerating scale.bandSize */
   scaleBandSize?: EnumSpecIndexTuple<number>[];
 
+  /** List of indice tuple for encoding mappings that require enumerating scale.domain */
+  scaleDomain?: EnumSpecIndexTuple<string | string[] | number[]>[];
```
Comment (kanitw) [inline src/model.ts]: string?

File: src/query/encoding.ts [hunk 5]
```diff
@@ -35,7 +35,9 @@ export interface BinQuery extends EnumSpec<boolean> {
 export interface ScaleQuery extends EnumSpec<boolean> {
   // TODO: add other properties from vegalite/src/scale
   clamp?: boolean | EnumSpec<boolean> | ShortEnumSpec;
+  domain?: string | number[] | string[] | EnumSpec<string | number[] | string[]> | ShortEnumSpec;
```
Comment (kanitw) [inline src/query/encoding.ts]: string

File: src/config.ts [hunk 6]
```diff
@@ -38,8 +38,22 @@ export interface QueryConfig {
 
   scaleBandSizes?: number[];
 
+  scaleClamps?: boolean[];
```
Comment (kanitw) [inline src/config.ts]: Why are you adding these `boolean[]`?

Comment (kanitw) [inline src/config.ts]: We never use them.

File: src/config.ts [hunk 7]
```diff
@@ -100,7 +114,15 @@ export const DEFAULT_QUERY_CONFIG: QueryConfig = {
...
 
   maxBinsList: [5, 10, 20],
+
   scaleBandSizes: [17, 21],
+
+  scaleDomains: [undefined],
+
+  scaleRanges: [undefined],
+
+  scaleExponents: [0.5, 1, 2],
```
Comment (kanitw) [inline src/config.ts]: [1]

File: src/query/encoding.ts [hunk 8]
```diff
@@ -34,8 +34,16 @@ export interface BinQuery extends EnumSpec<boolean> {
 
 export interface ScaleQuery extends EnumSpec<boolean> {
   // TODO: add other properties from vegalite/src/scale
+  clamp?: boolean | EnumSpec<boolean> | ShortEnumSpec;
+  domain?: string | number[] | string[] | EnumSpec<string | number[] | string[]> | ShortEnumSpec;
```
Comment (kanitw) [inline src/query/encoding.ts]: string ...

File: src/model.ts [hunk 9]
```diff
@@ -57,9 +57,30 @@ export interface EnumSpecIndex {
   /** List of indice tuple for encoding mappings that require enumerating scale.bandSize */
   scaleBandSize?: EnumSpecIndexTuple<number>[];
 
+  /** List of indice tuple for encoding mappings that require enumerating scale.clamp */
+  scaleClamp?: EnumSpecIndexTuple<boolean>[];
+
+  /** List of indice tuple for encoding mappings that require enumerating scale.domain */
+  scaleDomain?: EnumSpecIndexTuple<string | string[] | number[]>[];
```
Comment (kanitw) [inline src/model.ts]: string ...

**Pull request merged**